apis/Google.Cloud.Diagnostics.Common/Google.Cloud.Diagnostics.Common/SizedBufferingConsumer.cs

```diff
@@ -61,43 +63,115 @@ private SizedBufferingConsumer(IConsumer<T> consumer, ISizer<T> sizer, int buffe
         /// <param name="consumer">The consumer to flush to, cannot be null.</param>
         /// <param name="bufferSize">The buffer size in bytes.</param>
         public static SizedBufferingConsumer<T> Create(IConsumer<T> consumer, ISizer<T> sizer, int bufferSize)
+            => new SizedBufferingConsumer<T>(consumer, sizer, bufferSize);
+
+        /// <inheritdoc />
+        public void Receive(IEnumerable<T> items)
         {
-            return new SizedBufferingConsumer<T>(consumer, sizer, bufferSize);
+            GaxPreconditions.CheckNotNull(items, nameof(items));
+            try
+            {
+                _semaphore.Wait();
+                foreach (T item in items)
+                {
+                    _size += _sizer.GetSize(item);
+                    _items.Add(item);
+                    if (_size >= _bufferSize)
+                    {
+                        FlushesWithSemaphoreHeld();
+                    }
+                }
+            }
+            finally
+            {
+                _semaphore.Release();
+            }
         }
 
         /// <inheritdoc />
-        public void Receive(IEnumerable<T> items)
+        public async Task ReceiveAsync(
+            IEnumerable<T> items, CancellationToken cancellationToken = default(CancellationToken))
         {
             GaxPreconditions.CheckNotNull(items, nameof(items));
-            lock (_mutex)
+            try
             {
+                await _semaphore.WaitAsync(cancellationToken);
```

Comment:
Jon, is there a danger of the wait succeeding, but some exception being thrown between that line and the start of the try? If so, it might make sense to declare a bool outside, set to false, then use:
```csharp
waitSucceeded = await _semaphore.WaitAsync(-1, cancellationToken);
```
and then check that in the finally.

Although I guess you could make the same argument about something happening between WaitAsync returning and waitSucceeded getting set, so perhaps its not worth the added complexity.